Feat/45/missiles

grantramsay opened the pull request:
Partial fix for #45/#46 (see #45 for discussion).
Huge PR :O can split it up on request.
I'll add comments to PR on main bits that concerned me.
[Uploaded a short video here](https://gfycat.com/DescriptiveCraftyAgama).

To test click F5->F8 to cycle through active missiles, right click to fire/activate active missile.

grantramsay commented on apps/freeablo/farender/renderer.cpp:
```diff
@@ -277,9 +277,8 @@ namespace FARender
-                                  state->mPos.next().y,
-                                  state->mPos.getDist());
+                                  state->mPos.getFractionalPos().x,
+                                  state->mPos.getFractionalPos().y);
```
This would be cleaner (and simpler math in places) if `Misc::Point` stored `FixedPoint` coordinates rather than `int32_t`, i.e. currently the calculations have to be done in two parts (int and fractional), and two parameters passed (int and fractional) where it should be just one

grantramsay commented on apps/freeablo/faworld/missile/missile.h:
```diff
@@ -0,0 +1,165 @@
+        };
+
+        /******************** Missile classes ********************/
+        class MissileSingleFrame16Direction : public virtual Missile, public MissileMovementLinear, public MissileImpactDamageEnemyAndStop
```
Removed the need for this factory

grantramsay commented on apps/freeablo/faworld/missile/missilefactory.h:
```diff
@@ -0,0 +1,27 @@
+{
+    namespace Missile
+    {
+        // Xmacro craptory, there's probably a better way to do this while still minimising duplication.
```
Removed the need for this factory

wheybags commented:
This sounds really great, but it's not super easy to grok at the moment.
Could you write up a high-level overview of how the position and movement code works now, and also maybe a separate one on magic? Not just for me, I'd like to place them in the (underutilised) [docs folder](https://github.com/wheybags/freeablo/tree/master/docs).

As for merging, I would like to get the issues with floating point trig functions sorted out before we do merge, as I don't want to break mp/save-load now, when I'm planning on getting 0.4 out soon.

Also, you mentioned you can split this up on-request, which would be great. I'll let you decide what the best logical boundaries are, just smaller chunks to review is easier in general :p
Thanks very much! Super excited to see this shipping :D

grantramsay commented:
Ok will add some docs.

> As for merging, I would like to get the issues with floating point trig functions sorted out before we do merge, as I don't want to break mp/save-load now, when I'm planning on getting 0.4 out soon.
How would floating point trig functions break multiplayer save/load?
Fair enough if you don't want to risk breaking stuff for next release

Ok I can split it up, I might keep this one open for now for reference though.
I think each initial (non refactoring / bug fix) commit are split into pretty decent logical boundaries:
- fixed point fixes
- fixed point update
- direction update
- position update
- diablo.exe tables
- missiles
- hotkeys for testing

Also I want to change missile ownership from level to actor, as:
- spells like manashield need to follow the player across levels, so being bound to a level doesn't make sense
- for multiplayer if a player exits the game (or game is loaded without a player) their missiles now have a broken reference to a non-existent player. It would be better if their missiles, town portals etc just disappeared with them

grantramsay commented on apps/freeablo/faworld/position.cpp:
```diff
@@ -9,34 +9,67 @@ namespace FAWorld
+            int32_t disY = (FixedPoint::sin_degrees(mDirection.getDegrees() + 45) * vectorDist).round();
+
+            // TODO: Support non-trig (move diagonal at same speed) movement for actors (as in original game).
+            // Or was that just a bug in the original?
```
@wheybags what are your thoughts on this?

AJenbo commented:
> spells like manashield need to follow the player across levels, so being bound to a level doesn't make sense

ManaShild was originally lost when moving between levels, but other than that I agree :)

wheybags commented:
Thank you very much for this, and again, sorry for the prodigously long wait :v

wheybags commented:
> > spells like manashield need to follow the player across levels, so being bound to a level doesn't make sense
> 
> ManaShild was originally lost when moving between levels, but other than that I agree :)

Do you have a list of spells that should and shouldn't follow players across levels? If so, could you make a github issue here to implement that mechanic appropriately?

AJenbo commented:
> > > spells like manashield need to follow the player across levels, so being bound to a level doesn't make sense
> > 
> > 
> > ManaShild was originally lost when moving between levels, but other than that I agree :)
> 
> Do you have a list of spells that should and shouldn't follow players across levels? If so, could you make a github issue here to implement that mechanic appropriately?

None of them follow you across levels, most of there state logic is tied to there missiles (all spells have at least one missile, but some are invisible).
